pkg/insightstore/datapoint.go

```diff
@@ -0,0 +1,191 @@
+type DataPoint interface {
+	// Value gets data for model.InsightDataPoint
+	Value() float32
+	// Timestamp gets timestamp
+	GetTimestamp() int64
+}
+
+func Merge(dp1 DataPoint, dp2 DataPoint, kind model.InsightMetricsKind) (DataPoint, error) {
+	if dp1 == nil {
+		return dp2, nil
+	}
+	if dp2 == nil {
+		return dp1, nil
+	}
+
+	switch kind {
+	case model.InsightMetricsKind_DEPLOYMENT_FREQUENCY:
+		df1, ok := dp1.(DeployFrequency)
+		if !ok {
+			return nil, fmt.Errorf("cannot cast to DeployFrequency. DataPoint: %v", dp1)
+		}
+
+		df2, ok := dp2.(DeployFrequency)
+		if !ok {
+			return nil, fmt.Errorf("cannot cast to DeployFrequency. DataPoint: %v", dp1)
+		}
+
+		if df1.Timestamp != df2.Timestamp {
+			return nil, fmt.Errorf("mismatch timestamp. dp1: %d, dp2: %d", df1.Timestamp, df2.Timestamp)
+		}
+
+		df1.DeployCount += df2.DeployCount
+		return df1, nil
+	case model.InsightMetricsKind_CHANGE_FAILURE_RATE:
+		cfr1, ok := dp1.(ChangeFailureRate)
+		if !ok {
+			return nil, fmt.Errorf("cannot cast to ChangeFailureRate. DataPoint: %v", dp1)
+		}
+
+		cfr2, ok := dp2.(ChangeFailureRate)
+		if !ok {
+			return nil, fmt.Errorf("cannot cast to ChangeFailureRate. DataPoint: %v", dp2)
+		}
+
+		if cfr1.Timestamp != cfr2.Timestamp {
+			return nil, fmt.Errorf("mismatch timestamp. dp1: %d, dp2: %d", cfr1.Timestamp, cfr2.Timestamp)
+		}
+
+		cfr1.FailureCount += cfr2.FailureCount
+		cfr1.SuccessCount += cfr2.SuccessCount
+		cfr1.Rate = float32(cfr1.FailureCount) / float32(cfr1.FailureCount+cfr1.SuccessCount)
+		return cfr1, nil
+	default:
+		return nil, fmt.Errorf("invalid kind: %v", kind)
+	}
+}
+
+// convert types to list of DataPoint.
+func ToDataPoints(i interface{}) ([]DataPoint, error) {
+	switch dps := i.(type) {
+	case []DeployFrequency:
+		dataPoints := make([]DataPoint, len(dps))
+		for j, dp := range dps {
+			dataPoints[j] = dp
+		}
+		return dataPoints, nil
+	case []ChangeFailureRate:
+		dataPoints := make([]DataPoint, len(dps))
+		for j, dp := range dps {
+			dataPoints[j] = dp
+		}
+		return dataPoints, nil
+	default:
+		return nil, fmt.Errorf("cannot convert to DataPoints: %v", dps)
+	}
+}
+
+// AppendDataPoint append new data point to the end of list
+func AppendDataPoint(dp []DataPoint, point DataPoint) []DataPoint {
+	return append(dp, point)
+}
+
+// findDataPoint find key in the list of data points by timestamp
+func findDataPoint(dp []DataPoint, timestamp int64) (int, error) {
+	for i, d := range dp {
+		ts := d.GetTimestamp()
+		if ts == timestamp {
+			return i, nil
+		}
+	}
+	return 0, ErrNotFound
+}
+
+// GetDataPoint gets a data point by timestamp
+func GetDataPoint(dp []DataPoint, timestamp int64) (DataPoint, error) {
+	for _, d := range dp {
+		ts := d.GetTimestamp()
+		if ts == timestamp {
+			return d, nil
+		}
+	}
+	return nil, ErrNotFound
+}
+
+// SetDataPoint sets data point specified by timestamp
+func SetDataPoint(dp []DataPoint, point DataPoint, timestamp int64) []DataPoint {
+	k, err := findDataPoint(dp, timestamp)
```

Comment:
Aren't the data points sorted by the timestamp?
So we just compare it with the last one.

Reply:
As a specification of Go, there is a problem that when we iterate over a map with range, the order will change. Therefore, SetDataPoint is not always called in the sorted order, and datapoints during processing the loop are not guaranteed to be sorted.

It is true that having updatedps in an array instead of a map would guarantee that the order of insertion is sorted, and I feel it would be simpler, so I will make the change that way, thanks :D